cpp/pipemode/test/testPipeReader/TestPipeReader.cpp

```diff
@@ -0,0 +1,130 @@
+    EXPECT_EQ(reader.GetChannelDirectory(), channelDirectory);
+}
+
+TEST_F(PipeReaderTest, GetChannel) {
+    std::string channelDirectory = createTemporaryDirectory();
+    createChannel(channelDirectory, "elizabeth");
+    PipeReader reader(channelDirectory, "elizabeth");
+    EXPECT_EQ(reader.GetChannelName(), "elizabeth");
+}
+
+TEST_F(PipeReaderTest, ReadPipe) {
+    std::string channelDirectory = createTemporaryDirectory();
+    createChannel(channelDirectory, "elizabeth", "abc", 0);
+    PipeReader reader(channelDirectory, "elizabeth");
+    char buffer [4]; // 3 + 1 for the \0
+    size_t readNum = reader.Read(static_cast<void*>(buffer), 4);
+    EXPECT_STREQ("abc", buffer);
+    EXPECT_EQ(4, readNum);
+}
+
+TEST_F(PipeReaderTest, ReadPipeAfterEOF) {
+    std::string channelDirectory = createTemporaryDirectory();
+    createChannel(channelDirectory, "elizabeth", "abc", 0);
+    PipeReader reader(channelDirectory, "elizabeth");
+    char buffer [4];
+    size_t readNum = reader.Read(static_cast<void*>(buffer), 4);
+    readNum = reader.Read(static_cast<void*>(buffer), 1);
+    EXPECT_EQ(0, readNum);
+}
+
+TEST_F(PipeReaderTest, Reset) {
+    std::string channelDirectory = createTemporaryDirectory();
+    createChannel(channelDirectory, "elizabeth", "abc", 0);
+    PipeReader reader(channelDirectory, "elizabeth");
+    char buffer [4]; 
+    size_t readNum = reader.Read(static_cast<void*>(buffer), 4);
+    
+    // Create second iteration of elizabeth
+    createChannel(channelDirectory, "elizabeth", "def", 1);
+    reader.Reset();
+    readNum = reader.Read(static_cast<void*>(buffer), 4);
+    EXPECT_STREQ("def", buffer);
+    EXPECT_EQ(4, readNum);
+}
+
+TEST_F(PipeReaderTest, CloseReset) {
+    std::string channelDirectory = createTemporaryDirectory();
+    createChannel(channelDirectory, "elizabeth", "abc", 0);
+    PipeReader reader(channelDirectory, "elizabeth");
+    char buffer [4]; 
+    
+    // Read partially through the channel's current data
+    size_t readNum = reader.Read(static_cast<void*>(buffer), 2);
+    
+    // Close the current pipe
+    reader.Close();
+
+    // Create second iteration of elizabeth
+    createChannel(channelDirectory, "elizabeth", "def", 1);
+    reader.Reset();
+    readNum = reader.Read(static_cast<void*>(buffer), 4);
+    EXPECT_STREQ("def", buffer);
+    EXPECT_EQ(4, readNum);
+}
+
+TEST_F(PipeReaderTest, CloseImmediatelyReset) {
+    std::string channelDirectory = createTemporaryDirectory();
+    createChannel(channelDirectory, "elizabeth", "abc", 0);
+    PipeReader reader(channelDirectory, "elizabeth");
+    char buffer [4]; 
+    
+    // Close the current pipe, without reading anything from it
+    reader.Close();
+
+    // Create second iteration of elizabeth
+    createChannel(channelDirectory, "elizabeth", "def", 1);
+    reader.Reset();
+    size_t readNum = reader.Read(static_cast<void*>(buffer), 4);
+    EXPECT_STREQ("def", buffer);
+    EXPECT_EQ(4, readNum);
+}
```

Comment:
Can we have tests for the case where two readers are trying to read from the same stream?

Reply:
No, I'll add a comment saying that behavior is not supported. It won't happen in practice.